File: web/src/app/actions/strategies.ts
"use server";

import { getSession } from "@/app/actions/auth";
import * as api from "@/lib/api";

export type StrategyType =
  | "ema_crossover"
  | "sma_crossover"
  | "rsi_reversion"
  | "donchian_breakout";

export type StrategyFieldDefinition = {
  key: string;
  label: string;
  kind: "integer" | "decimal" | "boolean";
  description?: string | null;
  min?: string | null;
  max?: string | null;
  step?: string | null;
};

export type Strategy = {
  id: number;
  trading_account_id: number;
  name: string;
  strategy_type: StrategyType;
  ticker: string;
  symbols_json: string[];
  timeframe: "1Day";
  capital_allocation: string;
  params_json: Record<string, unknown>;
  risk_json: Record<string, unknown>;
  status: "active" | "paused" | "disabled";
  last_run_at: string | null;
  last_signal_at: string | null;
  last_error: string | null;
  created_at: string;
  updated_at: string;
};

export type StrategyRun = {
  id: number;
  strategy_id: number;
  trading_account_id: number;
  ticker: string;
  run_at: string;
  signal: "buy" | "sell" | "hold";
  action: "place_buy" | "place_sell" | "none";
  reason: string;
  inputs_json: Record<string, unknown>;
  order_id: number | null;
  error: string | null;
};

type StrategyListResponse = { strategies: Strategy[] };
type StrategyRunsResponse = {
  runs: StrategyRun[];
  total: number;
  page: number;
  per_page: number;
};

export type StrategyTemplate = {
  id: StrategyType;
  name: string;
  description: string;
  supported_timeframes: string[];
  default_params_json: Record<string, unknown>;
  default_risk_json: Record<string, unknown>;
  params_schema_json: StrategyFieldDefinition[];
  risk_schema_json: StrategyFieldDefinition[];
  status: string;
};

export type StrategyBacktestTrade = {
  ticker: string;
  side: string;
  quantity: string;
  price: string;
  timestamp: string;
  profit: string | null;
};

export type StrategyBacktestPoint = {
  time: number;
  equity: string;
  drawdown: string;
};

export type StrategyBacktestResult = {
  equity_curve: StrategyBacktestPoint[];
  drawdown_curve: StrategyBacktestPoint[];
  trades: StrategyBacktestTrade[];
  win_rate: number;
  avg_return_per_trade: number;
  max_drawdown: number;
  ending_equity: string;
};

export type StrategySnapshot = {
  trading_account_id: number;
  strategies: Strategy[];
  runs: StrategyRun[];
  open_orders: Array<Record<string, unknown>>;
  open_positions: Array<Record<string, unknown>>;
  strategy_executor_enabled: boolean;
};

export async function getStrategies(
  tradingAccountId: number,
): Promise<api.ApiResult<StrategyListResponse>> {
  const session = await getSession();
  if (!session) return { ok: false, error: "Not authenticated" };
  return api.get<StrategyListResponse>("/strategies", {
    trading_account_id: tradingAccountId.toString(),
  });
}

export async function createStrategy(payload: {
  trading_account_id: number;
  name: string;
  ticker: string;
  symbols_json?: string[];
  timeframe: "1Day";
  strategy_type?: StrategyType;
  status?: "active" | "paused" | "disabled";
  capital_allocation?: string;
  params_json: Record<string, unknown>;
  risk_json?: Record<string, unknown>;
}): Promise<api.ApiResult<Strategy>> {
  const session = await getSession();
  if (!session) return { ok: false, error: "Not authenticated" };
  return api.postJson<Strategy>("/strategies", payload);
}

export async function patchStrategy(
  strategyId: number,
  payload: {
    name?: string;
    ticker?: string;
    symbols_json?: string[];
    status?: "active" | "paused" | "disabled";
    timeframe?: "1Day";
    capital_allocation?: string;
    params_json?: Record<string, unknown>;
    risk_json?: Record<string, unknown>;
  },
): Promise<api.ApiResult<Strategy>> {
  const session = await getSession();
  if (!session) return { ok: false, error: "Not authenticated" };
  return api.patchJson<Strategy>(`/strategies/${strategyId}`, payload);
}

export async function deleteStrategy(strategyId: number): Promise<api.ApiResult<{ deleted: boolean }>> {
  const session = await getSession();
  if (!session) return { ok: false, error: "Not authenticated" };
  return api.del<{ deleted: boolean }>(`/strategies/${strategyId}`);
}

export async function runStrategy(strategyId: number): Promise<api.ApiResult<{ ok: boolean }>> {
  const session = await getSession();
  if (!session) return { ok: false, error: "Not authenticated" };
  return api.post<{ ok: boolean }>(`/strategies/${strategyId}/run`);
}

export async function getStrategyRuns(
  tradingAccountId: number,
  strategyId?: number,
): Promise<api.ApiResult<StrategyRunsResponse>> {
  const session = await getSession();
  if (!session) return { ok: false, error: "Not authenticated" };
  return api.get<StrategyRunsResponse>("/strategy-runs", {
    trading_account_id: tradingAccountId.toString(),
    strategy_id: strategyId?.toString(),
    page: "1",
    per_page: "50",
  });
}

export async function getStrategyCatalog(): Promise<api.ApiResult<{ templates: StrategyTemplate[] }>> {
  const session = await getSession();
  if (!session) return { ok: false, error: "Not authenticated" };
  return api.get<{ templates: StrategyTemplate[] }>("/strategy-catalog");
}

export async function getStrategySnapshot(
  tradingAccountId: number,
): Promise<api.ApiResult<StrategySnapshot>> {
  const session = await getSession();
  if (!session) return { ok: false, error: "Not authenticated" };
  return api.get<StrategySnapshot>("/strategy-snapshot", {
    trading_account_id: tradingAccountId.toString(),
  });
}

export async function runStrategyBacktest(payload: {
  strategy_type: StrategyType;
  ticker: string;
  symbols_json?: string[];
  timeframe: "1Day";
  capital_allocation: string;
  params_json: Record<string, unknown>;
  risk_json?: Record<string, unknown>;
  start: string;
  end: string;
}): Promise<api.ApiResult<StrategyBacktestResult>> {
  const session = await getSession();
  if (!session) return { ok: false, error: "Not authenticated" };
  return api.postJson<StrategyBacktestResult>("/strategies/backtest", payload);
}

export async function controlStrategies(payload: {
  trading_account_id: number;
  action: "pause_all" | "resume_all" | "disable_all";
}): Promise<api.ApiResult<{ updated: number; status: string }>> {
  const session = await getSession();
  if (!session) return { ok: false, error: "Not authenticated" };
  return api.postJson<{ updated: number; status: string }>("/strategy-controls", payload);
}

File: backend/app/schemas/strategies.py
from __future__ import annotations

from pydantic import BaseModel

from app.db.models import Strategy, StrategyRun


class StrategyResponse(BaseModel):
    id: int
    trading_account_id: int
    name: str
    strategy_type: str
    ticker: str
    symbols_json: list[str]
    timeframe: str
    capital_allocation: str
    params_json: dict
    risk_json: dict
    status: str
    last_run_at: str | None
    last_signal_at: str | None
    last_error: str | None
    created_at: str
    updated_at: str

    @classmethod
    def from_strategy(cls, strategy: Strategy) -> "StrategyResponse":
        return cls(
            id=strategy.id,
            trading_account_id=strategy.trading_account_id,
            name=strategy.name,
            strategy_type=strategy.strategy_type,
            ticker=strategy.ticker,
            symbols_json=list(strategy.symbols_json or []),
            timeframe=strategy.timeframe,
            capital_allocation=str(strategy.capital_allocation),
            params_json=strategy.params_json or {},
            risk_json=strategy.risk_json or {},
            status=strategy.status,
            last_run_at=strategy.last_run_at.isoformat()
            if strategy.last_run_at
            else None,
            last_signal_at=strategy.last_signal_at.isoformat()
            if strategy.last_signal_at
            else None,
            last_error=strategy.last_error,
            created_at=strategy.created_at.isoformat(),
            updated_at=strategy.updated_at.isoformat(),
        )


class StrategyListResponse(BaseModel):
    strategies: list[StrategyResponse]


class StrategyRunResponse(BaseModel):
    id: int
    strategy_id: int
    trading_account_id: int
    ticker: str
    run_at: str
    signal: str
    action: str
    reason: str
    inputs_json: dict
    order_id: int | None
    error: str | None

    @classmethod
    def from_run(cls, run: StrategyRun) -> "StrategyRunResponse":
        return cls(
            id=run.id,
            strategy_id=run.strategy_id,
            trading_account_id=run.trading_account_id,
            ticker=run.ticker,
            run_at=run.run_at.isoformat(),
            signal=run.signal,
            action=run.action,
            reason=run.reason,
            inputs_json=run.inputs_json or {},
            order_id=run.order_id,
            error=run.error,
        )


class StrategyRunsPageResponse(BaseModel):
    runs: list[StrategyRunResponse]
    total: int
    page: int
    per_page: int


class StrategyTemplateResponse(BaseModel):
    id: str
    name: str
    description: str
    supported_timeframes: list[str]
    default_params_json: dict
    default_risk_json: dict
    params_schema_json: list[dict]
    risk_schema_json: list[dict]
    status: str


class StrategyCatalogResponse(BaseModel):
    templates: list[StrategyTemplateResponse]


class StrategyBacktestTradeResponse(BaseModel):
    ticker: str
    side: str
    quantity: str
    price: str
    timestamp: str
    profit: str | None = None


class StrategyBacktestPointResponse(BaseModel):
    time: int
    equity: str
    drawdown: str


class StrategyBacktestResponse(BaseModel):
    equity_curve: list[StrategyBacktestPointResponse]
    drawdown_curve: list[StrategyBacktestPointResponse]
    trades: list[StrategyBacktestTradeResponse]
    win_rate: float
    avg_return_per_trade: float
    max_drawdown: float
    ending_equity: str


class StrategySnapshotResponse(BaseModel):
    trading_account_id: int
    strategies: list[StrategyResponse]
    runs: list[StrategyRunResponse]
    open_orders: list[dict]
    open_positions: list[dict]
    strategy_executor_enabled: bool
